Mage.Sets/src/mage/cards/w/WhirlwindDenial.java

```diff
@@ -55,36 +57,42 @@ public WhirlwindDenialEffect copy() {
+        // then pays that amount. Then each other opponent in turn order does the same.
+        // Then all spells and abilities that weren't paid for are countered at the same time.
+        // (2020-01-24)
+        Player opponent = game.getPlayer(game.getActivePlayerId());
+        do { // loop through all players starting with active to pay costs, but skip controller
+            if (game.getOpponents(source.getControllerId()).contains(opponent.getId())) {
+                for (StackObject stackObject : game.getStack()) {
+                    if (stackObject.getControllerId() != opponent.getId()) {
+                        continue;
                     }
-                    Player player = game.getPlayer(stackObject.getControllerId());
-                    if (player == null) {
-                        return;
-                    }
-                    Cost cost = ManaUtil.createManaCost(4, false);
-                    if (cost.canPay(source, source, stackObject.getControllerId(), game)
-                            && player.chooseUse(outcome, "Pay {4} to prevent "
+                    if (cost.canPay(source, source, opponent.getId(), game)
+                            && opponent.chooseUse(outcome, "Pay {4} to prevent "
                             + stackObject.getIdName() + " from being countered?", source, game)
                             && cost.pay(source, game, source, stackObject.getControllerId(), false)) {
-                        game.informPlayers("The cost was paid by "
-                                + player.getLogName()
-                                + " to prevent "
+                        game.informPlayers(opponent.getLogName()
+                                + " pays the cost to prevent "
                                 + stackObject.getIdName()
                                 + " from being countered.");
-                        return;
+                    } else {
+                        game.informPlayers(stackObject.getIdName()
+                                + " will be countered as "
+                                + opponent.getLogName()
+                                + " does not pay the cost.");
+                        stackObjectsToCounter.add(stackObject); // will be countered all at the end
                     }
-                    game.informPlayers("The cost was not paid by "
-                            + player.getLogName()
-                            + " to prevent "
-                            + stackObject.getIdName()
-                            + " from being countered.");
-                    game.getStack().counter(stackObject.getId(), source, game);
-                });
+                }
+            }
+            opponent = game.getPlayerList().getNextInRange(opponent, game); // move to next player
+        } while (opponent.getId() != game.getActivePlayerId()); // stop once looped through all players
```

Comment:
`opponent` can be null in some use cases (`getNextInRange` can return it), so must check it too. BUT I recommends to use `getPlayersInRange` code instead `while` loop. Infinite cycles must be used for infinite choose dialogs only like the auction logic cause it [can be buggy](https://github.com/magefree/mage/issues/5023) with AI or disconnected players:

Correct example:
``` java
        for (UUID playerId : game.getState().getPlayersInRange(controllerId, game)) {
            if (!playerId.equals(controllerId)) {
                // code here
            }
        }
```

All other looks fine.

Reply:
Thanks for the feedback. I updated with better logic per your recommendation.